staging/src/k8s.io/apiserver/pkg/endpoints/filters/impersonation/constrained_impersonation_test.go

```diff
@@ -0,0 +1,822 @@
+			requestInfo: &request.RequestInfo{
+				IsResourceRequest: true,
+				Verb:              "get",
+				APIVersion:        "v1",
+				Resource:          "pods",
+			},
+			expectedCode: http.StatusForbidden,
+		},
+		{
+			name: "allowed-legacy-impersonator",
+			user: &user.DefaultInfo{
+				Name: "legacy-impersonater",
+			},
+			expectedUser: &user.DefaultInfo{
+				Name:   "system:admin",
+				Groups: []string{"system:authenticated"},
+			},
+			expectedAttributes: []authorizer.AttributesRecord{
+				{
+					APIVersion: "v1",
+					Resource:   "pods",
+					Verb:       "impersonate-on:user-info:get",
+					User: &user.DefaultInfo{
+						Name: "legacy-impersonater",
+					},
+					ResourceRequest: true,
+				},
+				{
+					APIGroup:   authenticationapi.SchemeGroupVersion.Group,
+					APIVersion: authenticationapi.SchemeGroupVersion.Version,
+					Resource:   "users",
+					Name:       "system:admin",
+					Verb:       "impersonate:user-info",
+					User: &user.DefaultInfo{
+						Name: "legacy-impersonater",
+					},
+					ResourceRequest: true,
+				},
+				{
+					Resource: "users",
+					Verb:     "impersonate",
+					Name:     "system:admin",
+					User: &user.DefaultInfo{
+						Name: "legacy-impersonater",
+					},
+					ResourceRequest: true,
+				},
+			},
+			impersonationUser: "system:admin",
+			requestInfo: &request.RequestInfo{
+				IsResourceRequest: true,
+				Verb:              "get",
+				APIVersion:        "v1",
+				Resource:          "pods",
+			},
+			expectedCode: http.StatusOK,
+		},
+	}
+
+	var ctx context.Context
+	var actualUser user.Info
+	var lock sync.Mutex
+
+	doNothingHandler := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		currentCtx := req.Context()
+		user, exists := request.UserFrom(currentCtx)
+		if !exists {
+			actualUser = nil
+			return
+		}
+
+		actualUser = user
+
+		if _, ok := req.Header[authenticationapi.ImpersonateUserHeader]; ok {
+			t.Fatal("user header still present")
+		}
+		if _, ok := req.Header[authenticationapi.ImpersonateGroupHeader]; ok {
+			t.Fatal("group header still present")
+		}
+		for key := range req.Header {
+			if strings.HasPrefix(key, authenticationapi.ImpersonateUserExtraHeaderPrefix) {
+				t.Fatalf("extra header still present: %v", key)
+			}
+		}
+		if _, ok := req.Header[authenticationapi.ImpersonateUIDHeader]; ok {
+			t.Fatal("uid header still present")
+		}
+	})
+
+	constrainedAuthorizer := &constrainedImpersonateAuthorizer{}
+	handler := func(delegate http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+			defer func() {
+				if r := recover(); r != nil {
+					t.Errorf("Recovered %v", r)
+				}
+			}()
+
+			lock.Lock()
+			defer lock.Unlock()
+			req = req.WithContext(ctx)
+			currentCtx := req.Context()
+
+			user, exists := request.UserFrom(currentCtx)
+			if !exists {
+				actualUser = nil
+				return
+			} else {
+				actualUser = user
+			}
```

Comment:
Please refactor this to not abuse closures so much.  It is very hard to follow where these variables are being set up.  There should be no shared state between the test cases.